# Mnist 
Данный ноутбук используеться в учебных целях. 
В данной работе будут использоваться разные модели ML, но в основном - это модели классификации. 


Данной работой хочу показать своё умение работать с разными алгоритмами машинного обучения. А так же использовать его как своего рода шпаргалку. 
Здесь будут приведеные решения задачи из моего обучения.

In [1]:
import pandas as pd

from sklearn.decomposition import PCA # метод главных компонент, размерность
from sklearn.ensemble import AdaBoostClassifier # ada-бустинг, классификация
from sklearn.tree import DecisionTreeClassifier # решающее дерево, классификация
from sklearn.ensemble import RandomForestClassifier # случайный лес, классификация
from sklearn.ensemble import GradientBoostingClassifier # градиентный бустинг, классификация
from sklearn.svm import SVC # метод опорных векторов, классификация
from sklearn.linear_model import LogisticRegression # логистическая регрессия, классификация
from sklearn.ensemble import StackingClassifier # метод стекинга, обьедененеие несколькольких моделей, классификация

from sklearn.metrics import classification_report # отчет метрик для классификаторов
from sklearn.metrics import confusion_matrix # мастрица ошибок для классификации

Данные на которых будут обучаться модели - это данные mnist взятые с [kaggele](https://www.kaggle.com/datasets/oddrationale/mnist-in-csv/data). Они представляют из себя информацию о 60т образцах цифр, написанных от руки. Каждая строка состоит из 785 значений: первое значение (колонка label) — метка (число от 0 до 9), а оставшиеся 784 значения — значения пикселей (числа от 0 до 255)

Датасет состоит из двух частей: 

    обучающая mnist_train.csv - содержит 60т записей
    тестовая mnist_test.csv - содердит 10т записей

# Загрузка и обработка файлов

In [ ]:
# импортируем данные
df_train = pd.read_csv('C:/Users/Vabro/1workspace/dataset/ml_mnist/mnist_train.csv') # Загружаем тренировочный набор данных

Из данных отобрать изображение относящихся к 7 и 9.

In [3]:
# определяем переменные, которые которые будем отбирать
class_1 = 7 
class_2 = 9

df_train = df_train[(df_train['label']==class_1) | (df_train['label']==class_2)] # отбираем строки по классам
df_train, y_train = df_train.drop('label', axis=1), df_train['label'] # удаляем лишние данные из отфильтрованного датасета и сохроняем значения lable в отдельныю переменную.
df_train

# df_train содержит только данные изображений цифр
# y_train содержит метки классов


1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  1x10  ...  28x19  28x20  \
4        0    0    0    0    0    0    0    0    0     0  ...      0      0   
15       0    0    0    0    0    0    0    0    0     0  ...      0      0   
19       0    0    0    0    0    0    0    0    0     0  ...      0      0   
22       0    0    0    0    0    0    0    0    0     0  ...      0      0   
29       0    0    0    0    0    0    0    0    0     0  ...      0      0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  ...    ...    ...   
59973    0    0    0    0    0    0    0    0    0     0  ...      0      0   
59977    0    0    0    0    0    0    0    0    0     0  ...      0      0   
59988    0    0    0    0    0    0    0    0    0     0  ...      0      0   
59990    0    0    0    0    0    0    0    0    0     0  ...      0      0   
59992    0    0    0    0    0    0    0    0    0     0  ...      0      0   

       28x21  28x22  28x23  28x24  28x25  28x26  28x27  28x28  
4          0      0      0      0      0      0      0      0  
15         0      0      0      0      0      0      0      0  
19         0      0      0      0      0      0      0      0  
22         0      0      0      0      0      0      0      0  
29         0      0      0      0      0      0      0      0  
...      ...    ...    ...    ...    ...    ...    ...    ...  
59973      0      0      0      0      0      0      0      0  
59977      0      0      0      0      0      0      0      0  
59988      0      0      0      0      0      0      0      0  
59990      0      0      0      0      0      0      0      0  
59992      0      0      0      0      0      0      0      0  

[12214 rows x 784 columns]

In [4]:
df_test = pd.read_csv('C:/Users/Vabro/1workspace/dataset/ml_mnist/mnist_test.csv') # загружаем тестовые данные

df_test = df_test[(df_test['label']==class_1) | (df_test['label']==class_2)] # отбираем по условию
df_test, y_test = df_test.drop('label', axis=1), df_test['label'] # удаляем лишние данные из отфильтрованного датасета и сохроняем значения lable в отдельныю переменную
df_test
# df_test содержит только данные изображений цифр
# y_test содержит метки

1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  1x10  ...  28x19  28x20  \
0       0    0    0    0    0    0    0    0    0     0  ...      0      0   
7       0    0    0    0    0    0    0    0    0     0  ...      0      0   
9       0    0    0    0    0    0    0    0    0     0  ...      0      0   
12      0    0    0    0    0    0    0    0    0     0  ...      0      0   
16      0    0    0    0    0    0    0    0    0     0  ...      0      0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  ...    ...    ...   
9968    0    0    0    0    0    0    0    0    0     0  ...      0      0   
9973    0    0    0    0    0    0    0    0    0     0  ...      0      0   
9979    0    0    0    0    0    0    0    0    0     0  ...      0      0   
9990    0    0    0    0    0    0    0    0    0     0  ...      0      0   
9992    0    0    0    0    0    0    0    0    0     0  ...      0      0   

      28x21  28x22  28x23  28x24  28x25  28x26  28x27  28x28  
0         0      0      0      0      0      0      0      0  
7         0      0      0      0      0      0      0      0  
9         0      0      0      0      0      0      0      0  
12        0      0      0      0      0      0      0      0  
16        0      0      0      0      0      0      0      0  
...     ...    ...    ...    ...    ...    ...    ...    ...  
9968      0      0      0      0      0      0      0      0  
9973      0      0      0      0      0      0      0      0  
9979      0      0      0      0      0      0      0      0  
9990      0      0      0      0      0      0      0      0  
9992      0      0      0      0      0      0      0      0  

[2037 rows x 784 columns]

# Работа с моделями

## Метод главных компанент PCA

### Задание 1
Используйте метод главных компонент PCA() с параметром svd_solver='full'. Определите, какое минимальное количество главных компонент  необходимо использовать, чтобы доля объясненной дисперсии превышала 0.84 (была строго больше указанного значения) на тренировочном наборе данных.


In [5]:
pca_rate = 0.84 # устанавливаем дисперсию

pca = PCA(pca_rate, svd_solver='full') #создаем параметры для обучения
pca.fit(df_train) # обучаем модель на обучающих данных 

PCA(n_components=0.84, svd_solver='full')

In [6]:
df_pca_train = pca.transform(df_train) #применяем PCA к обучающим данным.

(12214, 42)

In [7]:
df_pca_test = pca.transform(df_test) # применяем PCA к тестовым данным.
df_pca_test.shape # смотрим размерноость

(2037, 42)

Минималькое кол-во главных компанент для уменьшение размерности, при требуемой доли сохроняемой дисперсии - 42.

### Задание 2
Найдите счеты, отвечающие найденным М главным компонентам для тренировочного и тестового наборов данных
координату относительно 6 главной компоненты для 10 изображения в тренировочном наборе данных

In [8]:
df_pca_train[9][5]

-242.79945649979683

## ADA boost

### Задание 3

Используйте полученные данные (два класса и сниженная размерность) для обучение нижеуказанных моделей.

На основе sklearn.ensemble.AdaBoostClassifier, используя DecisionTreeClassifier, обучите на тренировочных данных модель со следующими параметрами:
AdaBoostClassifier(
    base_estimator=DecisionTreeClassifier(max_depth=3, class_weight='balanced'),
    n_estimators=300,
    learning_rate=0.5,
    random_state=9)
    
Постройте classification_report и матрицу ошибок для полученной модели на тестовых данных

In [24]:
random_state = 9 # задаем случаныйе значения

# задаем параметры для обучения
ada_b = AdaBoostClassifier(
    estimator
    =DecisionTreeClassifier(max_depth=3,
                                          class_weight='balanced'), # Базовый классификатор - решающее дерево
    n_estimators=300, # Количество базовых классификаторов (деревьев) в ансамбле
    learning_rate=0.5,# Скорость обучения AdaBoost
    random_state=random_state) # Использование случайного значения для воспроизводимости

ada_b.fit(df_pca_train, y_train) # Обучение модель AdaBoost на тренировочных данных

AdaBoostClassifier(estimator=DecisionTreeClassifier(class_weight='balanced',
                                                    max_depth=3),
                   learning_rate=0.5, n_estimators=300, random_state=9)

In [10]:
y_pred = ada_b.predict(df_pca_test)# Использование ada_b для предсказания классов на тестовых данных df_pca_test
# Результаты сохраняются в y_pred
print(classification_report(y_test, y_pred, digits=4))
''' Вывод отчета о классификации,содержащего метрики качества
 (precision, recall, f1-score, support) для каждого класса
 и усредненные по классам значения
 digits=4 устанавливает количество знаков после запятой '''
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
print('TP: {}\nTN: {}\nFP: {}\nFN: {}'.format(tp, tn, fp, fn))

              precision    recall  f1-score   support

           7     0.9835    0.9835    0.9835      1028
           9     0.9832    0.9832    0.9832      1009

    accuracy                         0.9833      2037
   macro avg     0.9833    0.9833    0.9833      2037
weighted avg     0.9833    0.9833    0.9833      2037

TP: 992
TN: 1011
FP: 17
FN: 17


Рассмотрим метрики:

    Точность модели - 0.9833 
    True Positives - 992
    True Negatives - 1011
    False Positives - 17
    False Negatives - 17
    
В целом, модель хорошо себя показывает, но можно рассмотреть ложноположительные (FP) и ложноотрицательные (FN) отдельно, и постаратьсья уменьшить кол-во ошибок. 


## Градиентный бустинг

### Здание 4
Обучите модель на основе sklearn.ensemble.GradientBoostingClassifier
Базовые алгоритмы:(n_estimators=500, learning_rate=0.8, random_state=random_state, max_depth=2)

Постройте classification_report и матрицу ошибок для полученной модели на тестовых данных.

In [11]:
gb_clf = GradientBoostingClassifier(n_estimators=500, learning_rate=0.8, random_state=random_state, max_depth=2)
gb_clf.fit(df_pca_train, y_train)

GradientBoostingClassifier(learning_rate=0.8, max_depth=2, n_estimators=500,
                           random_state=11)

In [12]:
y_pred = gb_clf.predict(df_pca_test)
print(classification_report(y_test, y_pred, digits=4))
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
print('TP: {}\nTN: {}\nFP: {}\nFN: {}'.format(tp, tn, fp, fn))

              precision    recall  f1-score   support

           7     0.9852    0.9718    0.9785      1028
           9     0.9717    0.9851    0.9783      1009

    accuracy                         0.9784      2037
   macro avg     0.9784    0.9785    0.9784      2037
weighted avg     0.9785    0.9784    0.9784      2037

TP: 994
TN: 999
FP: 29
FN: 15


Рассмотрим метрики:

    Точность модели - 0.9784 
    True Positives - 994
    True Negatives - 999
    False Positives - 29
    False Negatives - 15
    
Для 9ки(или 9 класса) более низкая точнсть. Можно обратить внимание на данный момент. 


## Стакинг

### Задание 5

Обучите модель на основе sklearn.ensemble.StackingClassifier. Базовые алгоритмы: RandomForestClassifier(random_state=9), SVC(random_state=9). Решающий: LogisticRegression(random_state=9).

Постройте classification_report и матрицу ошибок для полученной модели на тестовых данных

In [32]:
base_estimators = [
    ('rf', RandomForestClassifier(random_state=random_state)), # указываем метод рандомного леса
    ('svc', SVC(random_state=random_state)) # указываем метод опорных векторов.
]

# Решающий классификатор
final_estimator = LogisticRegression(random_state=random_state)

# Создание StackingClassifier
stacking_clf = StackingClassifier(estimators=base_estimators, final_estimator=final_estimator)

In [22]:
stacking_clf.fit(df_pca_train, y_train) # обучаем модель на обучающих данных

StackingClassifier(estimators=[('rf', RandomForestClassifier(random_state=9)),
                               ('svc', SVC(random_state=9))],
                   final_estimator=LogisticRegression(random_state=9))

In [23]:
y_pred = stacking_clf.predict(df_pca_test) 
print(classification_report(y_test, y_pred, digits=4))
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
print('TP: {}\nTN: {}\nFP: {}\nFN: {}'.format(tp, tn, fp, fn))

              precision    recall  f1-score   support

           7     0.9961    0.9922    0.9942      1028
           9     0.9921    0.9960    0.9941      1009

    accuracy                         0.9941      2037
   macro avg     0.9941    0.9941    0.9941      2037
weighted avg     0.9941    0.9941    0.9941      2037

TP: 1005
TN: 1020
FP: 8
FN: 4


Рассмотрим метрики:

    Точность модели - 0.9941 
    True Positives - 1005
    True Negatives - 1020
    False Positives - 8
    False Negatives - 4
    
Отличная модель, добавить нечего.